src/lavinmq/mqtt/broker.cr

```diff
@@ -0,0 +1,111 @@
+
+      def initialize(@vhost : VHost, @replicator : Clustering::Replicator)
+        @sessions = Sessions.new(@vhost)
+        @clients = Hash(String, Client).new
+        @retain_store = RetainStore.new(File.join(@vhost.data_dir, "mqtt_reatined_store"), @replicator)
+        @exchange = MQTT::Exchange.new(@vhost, EXCHANGE, @retain_store)
+        @vhost.exchanges[EXCHANGE] = @exchange
+      end
+
+      def session_present?(client_id : String, clean_session) : Bool
+        return false if clean_session
+        session = sessions[client_id]?
+        return false if session.nil? || session.clean_session?
+        true
+      end
+
+      def add_client(socket, connection_info, user, packet)
+        if prev_client = @clients[packet.client_id]?
+          prev_client.close("New client #{connection_info.src} (username=#{packet.username}) connected as #{packet.client_id}")
+        end
+        client = MQTT::Client.new(socket,
+                                  connection_info,
+                                  user,
+                                  self,
+                                  packet.client_id,
+                                  packet.clean_session?,
+                                  packet.keepalive,
+                                  packet.will)
+        if session = sessions[client.client_id]?
+          if client.clean_session?
+            sessions.delete session
+          else
+            session.client = client
+          end
+        end
+        @clients[packet.client_id] = client
+        client
+      end
+
+      def remove_client(client)
+        client_id = client.client_id
+        if session = sessions[client_id]?
+          session.client = nil
+          sessions.delete(client_id) if session.clean_session?
+        end
+        @clients.delete client_id
+        vhost.rm_connection(client)
+      end
+
+      def publish(packet : MQTT::Publish)
+        @exchange.publish(packet)
+      end
+
+      def subscribe(client, packet)
+        unless session = sessions[client.client_id]?
+          session = sessions.declare(client.client_id, client.@clean_session)
+          session.client = client
+        end
+        qos = Array(MQTT::SubAck::ReturnCode).new(packet.topic_filters.size)
+        packet.topic_filters.each do |tf|
+          qos << MQTT::SubAck::ReturnCode.from_int(tf.qos)
+          session.subscribe(tf.topic, tf.qos)
+          @retain_store.each(tf.topic) do |topic, body|
+            headers = AMQP::Table.new
+            headers[RETAIN_HEADER] = true
+            msg = Message.new(EXCHANGE, topic, String.new(body),
+              AMQP::Properties.new(headers: headers,
+                delivery_mode: tf.qos))
+            session.publish(msg)
+          end
+        end
+        qos
+      end
+
+      def unsubscribe(client, packet)
+        session = sessions[client.client_id]
+        packet.topics.each do |tf|
+          session.unsubscribe(tf)
+        end
+      end
```

Comment:
nitpicking, but let the method know a little as possible of the world around it? but maybe for consitency you wanted the other way.. 

```suggestion
      def unsubscribe(client_id, topics)
        session = sessions[client_id]
        topics.each do |tf|
          session.unsubscribe(tf)
        end
      end
```